lib/kino/smart_cell.ex

```diff
@@ -160,6 +160,31 @@ defmodule Kino.SmartCell do
   """
   @callback to_source(attrs()) :: String.t()
 
+  @doc """
+  Invoked whenever the base evaluation context changes.
+
+  This callback receives the binding and environment available to the
+  smart cell code.
+
+  Note that this callback runs asynchronously and the result is then
+  handled with `c:scan_binding_continue/2` where the server state can
+  be updated accordingly.
+
+  **Important:** the return value is copied between processes, so
+  avoid returning large data structures. In particular, when looking
+  at variables, instead of returning their values, extract and return
+  only the relevant metadata.
+  """
+  @callback scan_binding(Code.binding(), Macro.Env.t()) :: data :: term()
+
+  @doc """
+  Invoked with the result of `c:scan_binding/2` to update the server
+  state.
+  """
+  @callback scan_binding_continue(data :: term(), ctx :: Context.t()) :: ctx :: Context.t()
```

Comment:
@josevalim we could similarly add `scan_evaluation_result` and `scan_evaluation_result_continue`. Implementation-wise I think it's pretty much the same as `handle_exception`. It's more verbose, but also gives more flexibility. Thoughts?

Reply:
To only handle errors we could then do:

```elixir
@impl true
def scan_evaluation_result({:ok, _}), do: :none
def scan_evaluation_result({:error, kind, reason, stacktrace}), do: {kind, reason, stacktrace}

@impl true
def scan_evaluation_result_continue(exception) do
  # handle_exception
end
```